### ISSUE TITLE
An unexpected result when using the pre-softmax layer instead of the softmax layer

### ISSUE THREAD

**Issue by hobincar** [chirag126/VOG#3]
Thank you for sharing the amazing work!

When I run the toy example, it runs perfectly fine and shows the exact same result you uploaded in this repo.
Below are the command and the result.
* command
   ```
   $ python3 toy_script.py --split test --mode early --vog_cal normalize
   ```
* result
   ![image](https://user-images.githubusercontent.com/17702664/176427800-6d057be9-504b-4fd4-93f3-6a7f48dc0d55.png)

However, I found that the code uses the gradient of the *softmax* layer w.r.t. the input, which differs from the paper in that the *pre-softmax* layer is used in the paper. So I changed [a single line of `toy_script.py`](https://github.com/chirag126/VOG/blob/master/toy_script.py#L175) as below and got a somewhat weird result when I run the code again.
* change
   ![Screenshot from 2022-06-29 20-53-18](https://user-images.githubusercontent.com/17702664/176430248-52174870-c9aa-41fe-b451-e6d855ba920a.png)
* result
  ![image](https://user-images.githubusercontent.com/17702664/176430375-02e26c6a-41e2-4745-b7d7-531039b494c8.png)

What did I miss here?

**Comment by dsouzadaniel** [comment]
Hey @hobincar ! Thanks for bringing this to our notice. This design is actually intentional for the toy problem : 

We backprop from the softmax layer because it is bounded between 0 and 1. However, if you look at the pre-softmax values, they are unbounded. Hence, when we backpropagate with **ones**, it can lead to unexpected (albeit interesting) results like this. 

On the contrary, the pre-softmax is the de-facto layer for backproping in the explainability field (Section 2 : https://arxiv.org/pdf/1312.6034.pdf ). Additionally, we see that pre-softmax values usually have better bounds due to techniques like BatchNorm, etc. 

I hope that answers the question! :)

**Comment by hobincar** [comment]
Thank you for the detailed explanation. Then when we choose the pre-softmax layer for backpropagating, which values would you recommend us to use to backpropagate with? I think the problem of using unbounded pre-softmax values would also appear with the ImageNet example.

**Comment by ddsouzapq1** [comment]
We use the pre-softmax itself. In all our experiments with MNIST/CIFAR-10/100/Imagenet, we've backpropagated from pre-softmax. It's just in this toy example we've used the post softmax. We've released the VoG scores for the MNIST/C-10/C-100 here : https://drive.google.com/uc?id=13on3KFgoTIh0pPKspTk6k-VJxBsZuRJy&export=download

**Comment by hobincar** [comment]
Thanks! It helps me a lot.
